electrum/base_wizard.py

```diff
@@ -368,9 +371,14 @@ def f(derivation, script_type):
                 self.run('on_hw_derivation', name, device_info, derivation, script_type)
             self.derivation_and_script_type_dialog(f)
         elif purpose == HWD_SETUP_DECRYPT_WALLET:
-            derivation = get_derivation_used_for_hw_device_encryption()
-            xpub = self.plugin.get_xpub(device_info.device.id_, derivation, 'standard', self)
-            password = keystore.Xpub.get_pubkey_from_xpub(xpub, ()).hex()
+            # The BitBox02 does not support non-standard xpubs
+            if device_info.label == "BitBox02":
+                xpub = self.plugin.get_xpub(device_info.device.id_, "m/84'/1'/0'", "p2wpkh", self)
+                password = hmac_oneshot(b"electrum_password_lulz", bitcoin.DecodeBase58Check(xpub), hashlib.sha512).hex()
+            else:
+                derivation = get_derivation_used_for_hw_device_encryption(device_info.label)
+                xpub = self.plugin.get_xpub(device_info.device.id_, derivation, 'standard', self)
+                password = keystore.Xpub.get_pubkey_from_xpub(xpub, ())
```

Comment:
Could we put the derivation path into the config and key each derivation path with the corresponding plugin name? Then we can get and set with the plugin name as argument and give the plugins the ability to override the default path.

The reason for the hmac is that I'm not feeling entirely comfortable with using a "normal" account level xpub for this operation. Maybe we could additionally store an argument to indicate that the derived xpub should be hashed.

Reply:
> Could we put the derivation path into the config 

The wallet files should be portable and work standalone. (e.g. user should be able to copy just a wallet file from one machine to another) (btw does this assumption break with the noise keys being in the config?)
Why would putting it into the config be better than just hardcoding it in the source?

> The reason for the hmac is that I'm not feeling entirely comfortable with using a "normal" account level xpub for this operation.

We have been using a custom hardened path exactly to avoid having to think about this stuff :) And also because the custom path is more likely to be "secret" (user might expose account xpub to e.g. accountant or maybe cosigner, but no reason to expose this custom path xpub).

I think the right way to go would be to introduce a new encrypted wallet file format, with new magic bytes, and a proper header (--we were already planning this with @ecdsa to be done later), and put this stuff into the header. At the same time we could do arbitrary changes to how the encryption works.
If we went this route, what should be in the header? for now, assume we keep hw-based file encryption restricted to non-multisig wallets. e.g. we could always do an hmac, and then put a salt and the bip32 path in the header

Reply:
> (btw does this assumption break with the noise keys being in the config?)

No, they will just be re-generated in the new wallet and the user will have to re-pair. This is expected behaviour, since every app should get its own static key.

> Why would putting it into the config be better than just hardcoding it in the source?

Special rules could be abstracted to the get and set functions instead of having them all over the place.

> (the) user should be able to copy just a wallet file from one machine to another

Ah, I see. Yeah, that breaks my config proposal.

> If we went this route, what should be in the header? 

This sounds very promising to me. From the top of my mind, how about adding as you said the salt and bip32 path, and additionally a version byte. Maybe @benma has some ideas for this as well.

Reply:
I've opened an issue for the storage file header: https://github.com/spesmilo/electrum/issues/5999

However, we've discussed this with @ecdsa, and that change (if done properly) is non-trivial, and it's not really a priority for us. We don't plan to do it for the next release (4.0) as we don't want to delay that any more.

Wallet files should be standalone as above, so putting it into the config file does not really work; and I would not like to maintain forever a hack that tries to multiple derivation paths.

Would it be possible to whitelist our custom hw encryption path in your firmware? It seems like small change. :)

Reply:
How about either

1) just moving the hack into the get_xpub() function of the BitBox02 plugin? 

```Python
def get_xpub(...):
    if keypath == get_derivation_used_for_hw_device_encryption():
        # return custom bitbox02 encryption key 
```

or b) add a hw plugin base method `get_encryption_key()` which every hw plugin can override?

In both cases, the workaround would be contained only to the bitbox02 plugin, not leaking into the base_wizard and elsewhere.

> Would it be possible to whitelist our custom hw encryption path in your firmware? It seems like small change. :)

We could do it as kind of a "last resort", but I don't think it is a nice solution for a few reasons:

1. it's mixing concerns - getting xpubs should be treated for what it is, allowing a client to generate receive addresses.
1. A special case for Electrum means technical debt that has to be treated differently. E.g. we have plans to have the user confirm exporting accounts (xpubs) on the device by name, so a special keypath like this would need special treatment. more special casing means more testing, more room for unforeseen issues, etc.
1. unclear if it can ever be removed? once wallet files are encrypted with this, it should be supported forever I assume.

Option `a)` above is a similar solution, but implemented in BitBox02 Electrum plugin instead of the actual firmware, so I'd go with that. 

Thoughts?

Reply:
Option a) would indeed be cleaner that your current proposal, and I could accept that. 
However, a separate derivation path for encryption would be cleaner than a hmac salted with "electrum_password_lulz".

In response to your point 2: Seriously? A special case for BitBox means technical debt for Electrum developers. We have no financial interest in accepting your plugin, the only thing we care about is long-term safety of our users. Please read https://github.com/spesmilo/electrum/blob/master/electrum/plugins/README

About point 3:  Once wallet are encrypted with a special derivation path, that special path must be supported forever. This is exactly the definition of technical debt. :-)

Reply:
(g1) Another option is that we could completely disable the storage (file) encryption option for bitbox02 wallets for now. This should not be much work. (note that in this case lightning keys would be plaintext on disk, when enabled)

(g2) With a bit more work, we could allow user-password based encryption for bitbox02 wallets. (This would not be wasted work as it would provide the bases for letting the user choose between userpw and xpubpw (https://github.com/spesmilo/electrum/issues/5561). I think it might not be a lot of work if we don't implement changing between userpw and xpubpw now.) However there might be unforeseen complications, so while I would take a look at this if it is liked, I cannot guarantee it would get implemented right now.

I would really rather not have bitbox02 use a different path for xpubpw. Currently we have the property that decrypting files only depends on your bip32 seed. So e.g. if you have a trezor that you used to encrypt a file, and then you lose the device and obtain a ledger initialised with the same seed, you can decrypt the file. Remember, the file might contain non-seed based stuff, such as tx history metadata and now even lightning state. This would still work with the header-based solution but not with the hacky workarounds.

(g3) Further note, re having an endpoint in the firmware to ask for our custom file enc password, that the current method does not need the full xpub, only the pubkey (not the chaincode, in particular). So you could conceivably create an endpoint that returns just the pubkey at exactly that bip32 path. This would then not interfere with your plans about prompting the user when an xpub is requested as without the chaincode this cannot be used as an account.

Reply:
Thanks all. My response landed here due to replying by email: https://github.com/spesmilo/electrum/pull/5993#issuecomment-592471118